File: frontends/nextjs/src/components/misc/demos/IRCWebchatDeclarative/functions/format-time.ts
import { useState, useEffect } from 'react'
import { useKV } from '@github/spark/hooks'
import type { User } from '@/lib/level-types'
import { getDeclarativeRenderer } from '@/lib/rendering-lib/declarative-component-renderer'
import { ChatWindow } from './irc/ChatWindow'
import { useChatInput, useFormattedTimes } from './irc/hooks'
import type { ChatMessage } from './irc/types'

export async function formatTime(timestamp: number): Promise<string> {
    try {
      const formatted = await renderer.executeLuaScript('lua_irc_format_time', [timestamp])
      return formatted || new Date(timestamp).toLocaleTimeString('en-US', { hour: '2-digit', minute: '2-digit' })
    } catch (error) {
      return new Date(timestamp).toLocaleTimeString('en-US', { hour: '2-digit', minute: '2-digit' })
    }
  }

File: packages/irc_webchat/seed/scripts/format_time.lua
---@param timestamp number Unix timestamp in milliseconds
---@return string Formatted time string in HH:MM AM/PM format
function formatTime(timestamp)
  local date = os.date("*t", timestamp / 1000)
  local hour = date.hour
  local ampm = "AM"

  if hour >= 12 then
    ampm = "PM"
    if hour > 12 then
      hour = hour - 12
    end
  end

  if hour == 0 then
    hour = 12
  end

  return string.format("%02d:%02d %s", hour, date.min, ampm)
end

return formatTime
